ibis/expr/datatypes.py

```diff
@@ -415,6 +415,32 @@ def valid_literal(self, value):
         return isinstance(value, six.string_types + (datetime.datetime,))
 
 
+class SignedInteger(Integer):
+    pass
+
+
+class UnsignedInteger(Integer):
+
+    @property
+    def bounds(self):
+        exp = self._nbytes * 8 - 1
+        upper = 1 << exp
+        return Bounds(lower=0, upper=upper)
+
+    def can_implicit_cast(self, other):
+        return (
+            isinstance(other, Integer) and
+            other >= 0 and
+            (type(self) is Integer or other._nbytes <= self._nbytes)
```

Comment:
`other` is not a value here, it's a type so `other >= 0` doesn't make sense here.

Reply:
Also, `type(self)` for `UnsignedInteger` instances will always be `False`.